### PR TITLE
Add XCNCIP2 ILS driver tests (and improve code)

### PR DESCRIPTION
**Pull request by xmorave2**
We are going to extend XCNCIP2 driver functionality, so I started by writing some tests. I'm not sure if anybody else is using this driver in production as original XC NCIP toolkit project seems abandoned, but we are using it a lot.

### CODE REVIEW

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: I notice that the original code used 'Accepted' for this condition, but you do not. Is this something that varies by instance? Should the list be configurable somewhere?

**Review by demiankatz — COMMENTED** [review]
This looks like a great start; I just have one question (see below).

I'm pretty sure that this is not being widely used in production, but it's always hard to be sure. A lot of the work here was done by @cedelis of CARLI, and they have moved away from VuFind due to a consortium-wide Alma/Primo contract, so I don't expect further development from there.

Regarding the overall status of the XCNCIP project, @mis306lu might have some insights....

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: It's a problem in NCIP standard itself - The values of fields are not defined - they are just strings... It depends only on implementation... If this driver i meant to work with XC NCIP toolkit based implementations, then I tried to fit to this library and it does not know the status "Accepted". That's why I changed this code. 

In fact we would need to make all these strings configurable... I's out of the scope of this PR for now I think...

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Makes sense to me! Just let me know when you would like to see this merged. Since these are primarily bug fixes, I could merge this now to get it included in the 7.0 release, but if you're planning to continue adding to this PR, I could also leave it open with the expectation that we'll make all the improvements at once in 7.1. I'm fine with either approach; just let me know what you prefer.

**Comment by xmorave2** [comment]
> I'm pretty sure that this is not being widely used in production, but it's always hard to be sure. A lot of the work here was done by @cedelis of CARLI, and they have moved away from VuFind due to a consortium-wide Alma/Primo contract, so I don't expect further development from there.

We are using NCIP, se there will be some development from our team I think....

> Regarding the overall status of the XCNCIP project, @mis306lu might have some insights....

I know only about folio plugin - https://github.com/folio-org/mod-ncip, which is based on xc ncip toolkit

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: I think it is  not necessary to have it in 7.0. I could continue to add some more tests, as these are only for LookupUser service. So, we can wait and merge later more work at once.

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Sounds good. I'll leave this open for now, and will await word from you when you are ready for further review and/or merging.

**Comment by xmorave2** [comment]
So I finished basic response and request tests and tried to fix all found problems. Could you make a review @demiankatz when you have some time? Thanks

File: languages/cs.ini
Comment (demiankatz) [inline languages/cs.ini]: Is this a generic correction/improvement that we should merge immediately, independent of the PR?

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: It might be slightly preferable to initialize this to `null` rather than leaving it uninitialized (just to make it clear that it's intentional).

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Since this isn't initialized by the constructor, it might be good to initialize it to null as well, so it has a clear default value if init() is not called.

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Since you're suppressing errors here, do you need to account for the possibility that simplexml_load_string fails and returns false?

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Should these status strings be held in a class property, and perhaps configurable via the .ini?

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: (I see that the same array is used again later in the code; might as well keep it in one place for maintainability).

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Is this something that needs to be addressed before the PR can be merged, or something for future consideration? Did the previous version of the code better address the consortial scenario with `$agencyList`?

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Would this be more concise as `(string)($holdCallNo[0] ?? '')`?

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Just to continue this conversation, and in light of my earlier comments about other configurable status messages, maybe it would make sense to create class properties for some of the statuses used here, just for consistency, and we can make them configurable in future as needed. It might also be good to be consistent about case-insensitive matching (which you did above, but not here).

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Strange that there are $underscored_variable_names here instead of $camelCasedVariableNames. Might be worth cleaning up for style consistency.

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: (I see that this inconsistency is found elsewhere in the class, though I haven't searched to see just how widespread it is. It's certainly not the most important thing to fix, so please don't feel pressured to waste lots of time on it... but I'm just mentioning it because I noticed it).

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: I see there's another instance of the same FIXME here; even if the problem can't be solved right away, is it possible that some centralization of logic can/should be done to have only one instance of the comment?

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php]: Can you avoid the is_string/is_array cases by simply checking for a non-empty value and then casting it to (array)?

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php:1279]: There's a missing doc comment here; missing comments in tests won't cause the build to fail, but it would probably be good to have for future reference.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for all of the work on this; it seems to be shaping up very nicely. I've left a few comments and questions for you.

File: languages/cs.ini
Comment (xmorave2) [reply languages/cs.ini]: It is generic correction, I just found it during this work..... So yes, it could be merged independently on this PR

**Comment by xmorave2** [comment]
Thanks for feedback @demiankatz. I am on vacation now, but I will work on this again after 6th of September.

**Comment by demiankatz** [comment]
Thanks, @xmorave2 -- I hope you have an enjoyable vacation!

File: languages/cs.ini
Comment (demiankatz) [reply languages/cs.ini]: Great, I've cherry-picked the relevant commit; one less file to worry about in this PR. :-)

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: fixed

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: fixed

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: It should not happen as `asXML` should return well-formed XML string, but things can always fail so I added a test on return value.

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: I added a properties and method for checking statuses strings

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Done, using regular expression it was not so much of work.

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Refactored

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php
Comment (xmorave2) [reply module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php]: Done

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php
Comment (xmorave2) [reply module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/XCNCIP2Test.php:1279]: Fixed

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Should this be case-insensitive and/or defined in a class property?

**Review by demiankatz — COMMENTED** [review]
Thanks for all the work on this, @xmorave2! I just happened to notice one other area in the code where there's a hard-coded status string that you might want to consider adjusting.

Beyond that, is there anything on your end that still needs to happen, or should I give this one last review with an eye to merging it?

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: Done. I even ended up with refactoring of getMyHolds and getMyStorageRetrievalRequests methods... ;)

**Comment by xmorave2** [comment]
> Beyond that, is there anything on your end that still needs to happen, or should I give this one last review with an eye to merging it?

It definitely will need some more improvements in the future, but for now I would to finish this and push it as is (after your final review).

File: module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php
Comment (xmorave2) [reply module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php]: I fix all these patterns (where) the default is empty string

**Comment by demiankatz** [comment]
@xmorave2, the latest changes look great. I've adjusted the comments in the .ini file slightly for clarity.

The only possible outstanding issue is that some of the properties and methods in the test class do not have comments on them; it might be worth adding those, but if you do not have time, it is not critical.

Please let me know whether or not you plan to make that final adjustment, and then I will merge after that if you have no other immediate changes to make.

Thanks again for all of the work on this!

**Comment by xmorave2** [comment]
@demiankatz: no problem, I added doc comment into test class. Thanks!

**Comment by demiankatz** [comment]
Thanks again, @xmorave2; I have merged this now. Further pull requests are welcome if you have additional changes to make!

**Pull request merged**
